File: scripts/categorize/fx_rates.py
"""Historical FX rate lookups via Frankfurter (free, no API key, ECB-based).
Falls back to a fixed approximate rate if the API is unreachable, so a
network hiccup doesn't take down the whole monthly import."""

import requests

FALLBACK_USD_TO_ILS = 2.97  # used only if the API call fails
_cache: dict[tuple[str, str], float] = {}


def get_rate(date: str, from_currency: str, to_currency: str = "ILS") -> float:
    """date is YYYY-MM-DD. Returns units of to_currency per 1 unit of from_currency."""
    key = (date, from_currency)
    if key in _cache:
        return _cache[key]

    try:
        response = requests.get(
            f"https://api.frankfurter.dev/v1/{date}",
            params={"from": from_currency, "to": to_currency},
            timeout=10,
        )
        response.raise_for_status()
        rate = response.json()["rates"][to_currency]
    except Exception as e:
        print(f"WARNING: FX rate lookup failed for {from_currency}->{to_currency} on {date} ({e}); using fallback rate {FALLBACK_USD_TO_ILS}")
        rate = FALLBACK_USD_TO_ILS

    _cache[key] = rate
    return rate

File: webapp/src/lib/fx-rates.ts
/** Historical FX rate lookups via Frankfurter (free, no API key, ECB-based).
 * Falls back to a fixed approximate rate if the API is unreachable, so a
 * network hiccup doesn't take down the whole monthly import. */

// Only used if the live API call fails. USD is the only currency we've
// actually seen in real data, so it's the only one with a real fallback
// number — anything else falls back to 1:1 with a loud warning rather than
// a made-up rate, since a fabricated "close enough" number for a currency
// we've never priced would be worse than an obviously-wrong 1:1.
const FALLBACK_RATES: Record<string, number> = {
  USD: 2.97,
};

const cache = new Map<string, number>();

/** date is YYYY-MM-DD. Returns units of toCurrency per 1 unit of fromCurrency. */
export async function getRate(date: string, fromCurrency: string, toCurrency = "ILS"): Promise<number> {
  const key = `${date}|${fromCurrency}|${toCurrency}`;
  const cached = cache.get(key);
  if (cached !== undefined) return cached;

  let rate: number;
  try {
    const response = await fetch(
      `https://api.frankfurter.dev/v1/${date}?from=${fromCurrency}&to=${toCurrency}`,
      { signal: AbortSignal.timeout(10_000) },
    );
    if (!response.ok) throw new Error(`HTTP ${response.status}`);
    const data = await response.json();
    rate = data.rates[toCurrency];
  } catch (e) {
    const fallback = FALLBACK_RATES[fromCurrency];
    if (fallback === undefined) {
      console.warn(
        `FX rate lookup failed for ${fromCurrency}->${toCurrency} on ${date} (${e}); no fallback rate known for ${fromCurrency}, using 1:1 — fix this amount manually`,
      );
      rate = 1.0;
    } else {
      console.warn(
        `FX rate lookup failed for ${fromCurrency}->${toCurrency} on ${date} (${e}); using fallback rate ${fallback}`,
      );
      rate = fallback;
    }
  }

  cache.set(key, rate);
  return rate;
}
